ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java

```diff
@@ -135,6 +152,177 @@ public static void updateCrosslinks(BeaconState state) throws Exception {
+      case_one_penalties_and_rewards(
+          state, balances, previous_total_balance, previous_balance, previous_indices);
+
+      // Expected FFG target
+      previous_balance = AttestationUtil.get_previous_epoch_boundary_attesting_balance(state);
+      previous_indices = AttestationUtil.get_previous_epoch_boundary_attester_indices(state);
+      case_one_penalties_and_rewards(
+          state, balances, previous_total_balance, previous_balance, previous_indices);
+
+      // Expected beacon chain head
+      previous_balance = AttestationUtil.get_previous_epoch_head_attesting_balance(state);
+      previous_indices = AttestationUtil.get_previous_epoch_head_attester_indices(state);
+      case_one_penalties_and_rewards(
+          state, balances, previous_total_balance, previous_balance, previous_indices);
+
+      // Inclusion distance
+      UnsignedLong reward_delta = UnsignedLong.ZERO;
+      previous_indices = AttestationUtil.get_previous_epoch_attester_indices(state);
+      for (int index : previous_indices) {
+        UnsignedLong inclusion_distance = AttestationUtil.inclusion_distance(state, index);
+        reward_delta =
+            base_reward(state, index, previous_total_balance)
+                .times(UnsignedLong.valueOf(MIN_ATTESTATION_INCLUSION_DELAY))
+                .dividedBy(inclusion_distance);
+        apply_penalty_or_reward(balances, index, reward_delta, true);
+      }
+
+      // Case 2: epochs_since_finality > 4:
+    } else if (epochs_since_finality.compareTo(FOUR) > 0) {
+
+      Function<Integer, UnsignedLong> inactivity_penalty =
+          new Function<Integer, UnsignedLong>() {
+            @Override
+            public UnsignedLong apply(Integer index) {
+              return inactivity_penality(
+                  state, index, epochs_since_finality, previous_total_balance);
+            }
+          };
+
+      Function<Integer, UnsignedLong> base_penalty =
+          new Function<Integer, UnsignedLong>() {
+            @Override
+            public UnsignedLong apply(Integer index) {
+              return base_reward(state, index, previous_total_balance);
+            }
+          };
+
+      // prev epoch justified attester
+      List<Integer> validator_indices =
+          AttestationUtil.get_previous_epoch_justified_attester_indices(state);
+      case_two_penalties_and_rewards(state, balances, validator_indices, inactivity_penalty);
+      // prev epoch boundary attester
+      validator_indices = AttestationUtil.get_previous_epoch_boundary_attester_indices(state);
+      case_two_penalties_and_rewards(state, balances, validator_indices, inactivity_penalty);
+      // prev epoch head attester
+      validator_indices = AttestationUtil.get_previous_epoch_head_attester_indices(state);
+      case_two_penalties_and_rewards(state, balances, validator_indices, base_penalty);
+
+      Function<Integer, UnsignedLong> composite_penalty =
+          new Function<Integer, UnsignedLong>() {
```

Comment:
how do you define composite_penalty as a function with two parameters, but then when implementing it inside, you just have one variable, the index integer?

Reply:
Integer is the incoming value and Unsigned long is the outgoing variable.